### PR TITLE
[Parse] Improve diagnostic for deprecated protocol<...> syntax

### PR DESCRIPTION
**Pull request by rintaro**
#### What's in this pull request?
- Don't emit deprecated warnings for incomplete protocol<...>.
  
  ```
  func fn(x: protocol<P) {}
  ```
  
  Because other _error_ diagnostics are already emitted
  Moreover, the current fix-it tries to remove `)` after `P`
- Different message for single protocol compostion.
  
  ```
  typealias MyError = protocol<Error>
  ```
  
  Because `"join the protocols using '&'"` is not relevant in this case

CC: @DougGregor, @joewillsher 

---

Before merging this pull request to apple/swift repository:
- [ ] Test pull request on Swift continuous integration.

<details>
  <summary>

Triggering Swift CI</summary>

The swift-ci is triggered by writing a comment on this PR addressed to the GitHub user @swift-ci. Different tests will run depending on the specific comment that you use. The currently available comments are:

**Smoke Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please smoke test |
| All supported platforms | @swift-ci Please smoke test and merge |
| OS X platform | @swift-ci Please smoke test OS X platform |
| Linux platform | @swift-ci Please smoke test Linux platform |

A smoke test on macOS does the following:
1. Builds the compiler incrementally.
2. Builds the standard library only for macOS. Simulator standard libraries and
   device standard libraries are not built.
3. lldb is not built.
4. The test and validation-test targets are run only for macOS. The optimized
   version of these tests are not run.

A smoke test on Linux does the following:
1. Builds the compiler incrementally.
2. Builds the standard library incrementally.
3. lldb is build incrementally.
4. The swift test and validation-test targets are run. The optimized version of these
   tests are not run.
5. lldb is tested.

**Validation Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please test |
| All supported platforms | @swift-ci Please test and merge |
| OS X platform | @swift-ci Please test OS X platform |
| OS X platform | @swift-ci Please benchmark |
| Linux platform | @swift-ci Please test Linux platform |

**Lint Testing**

| Language | Comment |
| --- | --- |
| Python | @swift-ci Please Python lint |

Note: Only members of the Apple organization can trigger swift-ci.
</details>

### CODE REVIEW

**Comment by joewillsher** [comment]
This looks great, I think maybe the error message should specify that the syntax ‘is deprecated and *not needed here*’ rather than implying the construct is useless?

File: lib/Parse/ParseType.cpp [hunk 1]
```diff
@@ -458,17 +458,23 @@ ParserResult<TypeRepr> Parser::parseTypeIdentifierOrTypeComposition() {
...
       Context, Protocols, ProtocolLoc, {LAngleLoc, RAngleLoc});
 
-    Diag.highlight({ProtocolLoc, RAngleLoc});
-    for (auto Comma : Commas)          // remove commas and '<'
-      Diag.fixItReplace(Comma, " &");
-    Diag
-      .fixItRemove({ProtocolLoc, LAngleLoc}) // remove 'protocol<'
-      .fixItRemove(RAngleLoc); // remove '>'
+    if (Status.isSuccess()) {
+      // Only if we have complete prtocol<...> composition, diagnose deprecated.
```
Comment (joewillsher) [inline lib/Parse/ParseType.cpp]: Small nitpick: typo in this comment

**Comment by rintaro** [comment]
@joewillsher Thank you for the suggestion! I'll do that.
Sorry, I'm not good at English 😓

**Comment by joewillsher** [comment]
No problem at all, thanks doing this!

**Comment by rintaro** [comment]
Updated warning message for single protocol composition: `protocol<P>`

File: test/expr/expressions.swift [hunk 2]
```diff
@@ -230,7 +230,7 @@ func test_lambda() {
 
 func test_lambda2() {
   { () -> protocol<Int> in
-    // expected-warning @-1 {{'protocol<...>' composition syntax is deprecated; join the protocols using '&'}}
+    // expected-warning @-1 {{'protocol<...>' composition syntax is deprecated and not needed here}}
```
Comment (CodaFi) [inline test/expr/expressions.swift]: Could you please test the fixit is working here?

File: test/type/protocol_composition.swift [hunk 3]
```diff
@@ -123,10 +123,9 @@ func testConversion() {
 }
 
 // Test the parser's splitting of >= into > and =.
-var x : protocol<P5>= 17 // expected-warning {{'protocol<...>' composition syntax is deprecated; join the protocols using '&'}}
+var x : protocol<P5>= 17 // expected-warning {{'protocol<...>' composition syntax is deprecated and not needed here}}
```
Comment (CodaFi) [inline test/type/protocol_composition.swift]: Same here, just like you did in the bottom of this diff.

File: test/expr/expressions.swift (hunk 2 above)
Comment (rintaro) [inline test/expr/expressions.swift]: Thank you! done.

**Comment by CodaFi** [comment]
Thanks all!

@swift-ci please test and merge.

**Comment by rintaro** [comment]
@swift-ci Please smoke test

**Comment by rintaro** [comment]
@swift-ci  Please test

**Comment by rintaro** [comment]
@swift-ci Please test

**Comment by rintaro** [comment]
Tests passed, merging.

**Comment by DougGregor** [comment]
Looks great, thank you!

**Pull request merged**